server/src/main/java/org/opensearch/indices/IndicesService.java

```diff
@@ -409,7 +409,15 @@ public IndicesService(
         this.shardsClosedTimeout = settings.getAsTime(INDICES_SHARDS_CLOSED_TIMEOUT, new TimeValue(1, TimeUnit.DAYS));
         this.analysisRegistry = analysisRegistry;
         this.indexNameExpressionResolver = indexNameExpressionResolver;
-        this.indicesRequestCache = new IndicesRequestCache(settings);
+        this.indicesRequestCache = new IndicesRequestCache(settings, (shardId -> {
+            IndexService indexService = null;
+            try {
+                indexService = indexServiceSafe(shardId.getIndex());
+            } catch (IndexNotFoundException ex) {
+                return Optional.empty();
+            }
```

Comment:
This feels like a slight abuse of `indexServiceSafe`.

Can you use `this.IndicesService.indices.get()` instead, and return `empty` on `null`? That should capture the `IndicesService` instance in the lambda, which will give you the current `indices` reference at the time the lambda is invoked (rather than accidentally capturing `indices` in the constructor).

Reply:
Sure. Changed.